rpc_util.go

```diff
@@ -817,17 +817,70 @@ func decompress(compressor encoding.Compressor, d []byte, maxReceiveMessageSize
 			//
 			// TODO: If we ensure that the buffer size is the same as the DecompressedSize,
 			// we can also utilize the recv buffer pool here.
-			buf := bytes.NewBuffer(make([]byte, 0, size+bytes.MinRead))
-			bytesRead, err := buf.ReadFrom(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1))
+			bufferSize := uint64(size) + bytes.MinRead
+			if bufferSize > math.MaxInt {
+				bufferSize = math.MaxInt
+			}
+			buf := bytes.NewBuffer(make([]byte, 0, int(bufferSize)))
+			bytesRead, err := buf.ReadFrom(io.LimitReader(dcReader, int64(maxReceiveMessageSize)))
+			if err != nil {
+				return nil, int(bytesRead), err
+			}
+
+			if err = checkReceiveMessageOverflow(bytesRead, int64(maxReceiveMessageSize), dcReader); err != nil {
+				return nil, size + 1, err
+			}
 			return buf.Bytes(), int(bytesRead), err
+
 		}
 	}
-	// Read from LimitReader with limit max+1. So if the underlying
-	// reader is over limit, the result will be bigger than max.
-	d, err = io.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)+1))
+	// This function reads data from the `dcReader` stream, limiting the amount of data read to the
+	// `maxReceiveMessageSize` using `io.LimitReader`. The `io.ReadAll` function is used to read
+	// all available data within the specified limit.
+	//
+	// If an error occurs during the read operation, the function returns `nil`, the length of the data
+	// read so far, and the error.
+	//
+	// Returns:
+	//   - d: The data read from the stream, or nil if an error occurs.
+	//   - int: The length of the data read.
+	//   - error: An error if the read operation fails, or nil if the operation is successful.
+	d, err = io.ReadAll(io.LimitReader(dcReader, int64(maxReceiveMessageSize)))
+	if err != nil {
+		return nil, len(d), err
+	}
+	if err = checkReceiveMessageOverflow(int64(len(d)), int64(maxReceiveMessageSize), dcReader); err != nil {
+		return nil, len(d) + 1, err
+	}
 	return d, len(d), err
 }
 
+// checkReceiveMessageOverflow checks if the number of bytes read from the stream exceeds
+// the maximum receive message size allowed by the client. If the `readBytes` equals
+// `maxReceiveMessageSize`, the function attempts to read one more byte from the `dcReader`
+// to detect if there's an overflow.
+//
+// If additional data is read, or an error other than `io.EOF` is encountered, the function
+// returns an error indicating that the message size has exceeded the permissible limit.
+//
+// Parameters:
```

Comment:
this is too verbose. You don't have to write params and return for docstring. It is clear from the function signature